iroha_http_server/src/lib.rs

```diff
@@ -0,0 +1,640 @@
+//! HTTP/1.1 server library with WebSocket support heavily inspired by [tide](https://crates.io/crates/tide).
+//TODO: do we need TLS/SSL?
+use async_std::{
+    net::{TcpListener, TcpStream},
+    prelude::*,
+    task,
+};
+use async_trait::async_trait;
+pub use async_tungstenite::tungstenite::Message as WebSocketMessage;
+use async_tungstenite::WebSocketStream as TungsteniteWebSocketStream;
+use httparse::{Request as HttpParseRequest, Status};
+use route_recognizer::Router;
+use std::{
+    borrow::Borrow,
+    collections::BTreeMap,
+    convert::{TryFrom, TryInto},
+    fmt::Display,
+    sync::Arc,
+};
+
+pub const GET_METHOD: &str = "GET";
+pub const POST_METHOD: &str = "POST";
+pub const WEB_SOCKET_UPGRADE: &[u8] = b"websocket";
+pub const ALLOW_HEADER: &str = "Allow";
+pub const UPGRADE_HEADER: &str = "Upgrade";
+pub const CONTENT_LENGTH_HEADER: &str = "Content-Length";
+pub const HTTP_CODE_OK: u16 = 200;
+pub const HTTP_CODE_INTERNAL_SERVER_ERROR: u16 = 500;
+pub const HTTP_CODE_NOT_FOUND: u16 = 404;
+pub const HTTP_CODE_BAD_REQUEST: u16 = 400;
+pub const HTTP_CODE_METHOD_NOT_ALLOWED: u16 = 405;
+pub const HTTP_VERSION_1_1: &str = "HTTP/1.1";
+const MAX_HEADERS: usize = 128;
+const BUFFER_SIZE: usize = 4096;
+
+pub struct Server<State> {
+    state: State,
+    router: Arc<Router<Endpoint<State>>>,
+}
+
+impl<State: Clone + Send + Sync + 'static> Server<State> {
+    pub fn new(state: State) -> Self {
+        Server {
+            state,
+            router: Arc::new(Router::new()),
+        }
+    }
+
+    pub fn at<'s>(&'s mut self, path: &str) -> RouteBuilder<'s, State> {
```

Comment:
Frustrating naming in my opinion, `route` may be better.

Reply:
Used this naming from `tide`. Not a problem to change it, but it just creates a nice flow of `server.at(url).get(handler)`